text/0000-codify-infrastructure.md

```diff
@@ -0,0 +1,392 @@
+    + [Pods 🐟](#pods-%F0%9F%90%9F)
+    + [Services 🚃](#services-%F0%9F%9A%83)
+    + [Ingress 🚪](#ingress-%F0%9F%9A%AA)
+    + [Nodes 🧱](#nodes-%F0%9F%A7%B1)
+  * [Proposed Cluster Architecture 🏗](#proposed-cluster-architecture-%F0%9F%8F%97)
+    + [Listeners (fruitcakes) 🎧](#listeners-fruitcakes-%F0%9F%8E%A7)
+    + [DJs and VJs 💽](#djs-and-vjs-%F0%9F%92%BD)
+    + [Developers 👩🏻‍💻](#developers-%F0%9F%91%A9%F0%9F%8F%BB%E2%80%8D%F0%9F%92%BB)
+    + [Management 🤵🏻](#management-%F0%9F%A4%B5%F0%9F%8F%BB)
+  * [Path to Production 👷🏻‍♀️](#path-to-production-%F0%9F%91%B7%F0%9F%8F%BB%E2%80%8D%E2%99%80%EF%B8%8F)
+- [Drawbacks 😬](#drawbacks-%F0%9F%98%AC)
+- [Alternatives ⚖️](#alternatives-%E2%9A%96%EF%B8%8F)
+- [Unresolved questions ❓](#unresolved-questions-%E2%9D%93)
+
+<!-- tocstop -->
+
+## Summary 📖
+
+Right now, the infrastructure for datafruits is really ad-hoc. As best as I can tell, some stuff is running on [Heroku](https://www.heroku.com/) and some stuff is running on [DigitalOcean](https://www.digitalocean.com/). This RFC has two goals:
+
+1. Codify the infrastructure for the datafruits site so that it's easily understandable where everything lives.
+2. Cloud-ify the existing infrastructure so that each component is individually scalable.
+
+## Motivation 🏃🏻‍♀️
+
+As stated above, one reason to do this is to make it more clear where everything lives -- it allows someone (like me ☺️) to take a look and understand how all the pieces fit together.
+
+Another reason to keep the infrastructure as code is that it makes the infra versionable, meaning rollbacks are literally as easy as `git revert`. [More on the benefits of IoC](https://docs.microsoft.com/en-us/azure/devops/learn/what-is-infrastructure-as-code)
+
+This also allows us to move from treating the services (streampusher, datafruits) like pets to treating them like cattle, which means less messy ssh-ing and more codified changes. [More reading on this terminology](https://medium.com/@Joachim8675309/devops-concepts-pets-vs-cattle-2380b5aab313)
+
+Additionally, by moving to a [Kubernetes (K8s)](https://kubernetes.io/) based infrastructure, it allows us a lot more (horizontal) scalability, and the ability to scale out each chunk of the stack individually -- if there are 30,000,000 users in the chat, it makes sense to scale up the chat server, but maybe not also the [rmtp(s)](https://en.wikipedia.org/wiki/Real-Time_Messaging_Protocol) server, which only has the one person running viz connected to it.
+
+Fourth, we now have the opportunity to actually do logging. Logging! Imagine, datafruits goes down and we don't have to spend eight hours figuring out why. We have an idea for a new feature, and we can actually figure out if people use it. Metrics! God, the metrics. Plus we can avoid G\*\*\*\*\* A\*\*\*\*\*\*\*\* and run our [own analytics platform](https://github.com/zgoat/goatcounter) that's privacy friendly.
```

Comment:
Can you elaborate more on how the logging would work? Is there some kind of web interface? I'm new to this...

Reply:
Yeah sure! Distributed logging is actually a kind of big field, because everything is so spread out (heh). Here's a [quick explanation](https://www.reddit.com/r/devops/comments/9hku3v/prometheus_vs_opentracing/e6db02o) of "observability", which is CS-speak for understanding what's going on inside a distributed system. Copied here for your convenience:

> There are 3 pillars of observability:
> 
> * metrics
> * logs
> * tracing
> 
> While all of them partially overlap, each have different purpose.
>
> #### Metrics
> 
> These are numbers related amount of events that happened in the time range. Like:
> 
> * number of successful/failed/overall requests
> * cumulative duration of requests
> * bucketed histogram of requests’ durations
> 
> The main point of metrics is that these are small, you can gather them cheaply and store them for a long period of time. These give you overall overview of the whole system, but without insights.
> 
> So, metrics answer to the question “how is my system performance changes through time?”
>
> #### Logs
> 
> These are single events that happened in the system, ex. single request. It will often contain also some informations that are in the metrics (like request duration), but it will also contain more context, like IP of the requester and precise time of the request. The problem with logs isn’t that these are larger in size than just metrics itself, so you cannot store them for so long as the metrics and due tho their size, you need to reduce amount of logs you send.
> 
> So, logs answers to the question “what happened in my system?”.
>
> #### Tracing
> 
> This one is for monitoring how single event is behaving in the system. So you have information how long this request spent in LB, backend, DB, how long it spend in external services, and so on. So as you see, this overlap with both of metrics and logs, but still not exactly. The main problem is that while metrics and logs can have different formats in each of the services, tracing needs to be uniform, because your system needs to be “request ID aware”. Also traces can be very large, that is why in general you do not trace all requests, but only some of them to reduce internal traffic.
> 
> So, tracing answers to the question “how is my system components interacting with each other?”.

#### Cluster-Level Logging

This is like, the classic version of logging. You have an application, and it spits some stuff out via `console.log` or `console.error`, and you want to capture that. In Kubernetes, every container's `stdout` and `stderr` is captured, and you have the option of monitoring logs either at the Pod (container group) level or at the Node (host VM) level.

We'd almost certainly do it at the Node level, because that doesn't add any additional overhead to application programming. Here's a nice image [from the docs](https://kubernetes.io/docs/concepts/cluster-administration/logging/#using-a-node-logging-agent):

![node logging](https://d33wubrfki0l68.cloudfront.net/2585cf9757d316b9030cf36d6a4e6b8ea7eedf5a/1509f/images/docs/user-guide/logging/logging-with-node-agent.png)

What you're looking at is a simple logging configuration for one node. You have your application container, which streams to `stdout` and `stderr`, which is then redirected by Kubernetes to a log file on the Node. The "logging agent", which is actually called [Fluentd](https://docs.fluentd.org/), parses these logs and sends them off to some other log aggregation service. `logrotate` is a built-in Kubernetes tool that just deletes old logs once they take up a certain amount of disk space.

#### Application-Level Logging

This is a little more involved than just streaming to `stdout` and `stderr`, but not much more so. At least, on the application side. Usually what you do is use something like a [Prometheus client library](https://github.com/deadtrickster/prometheus.ex) and write something like this:

```ex
defmodule ExampleInstrumenter do
  use Prometheus.Metric

  @histogram [name: :http_request_duration_milliseconds,
              labels: [:method],
              buckets: [100, 300, 500, 750, 1000],
              help: "Http Request execution time"]

  def instrument(%{time: time, method: method}) do
    Histogram.observe([name: :http_request_duration_milliseconds, labels: [method]], time)
  end
end
```

This is useful to monitor metrics from within individual applications, such as response times on the API or which DJs are the most obsessive about updating their DF profiles. It's basically strictly better than cluster-level logging, except that it requires you to actually write code dedicated to instrumenting logging instead of just `console.log`-ing.

#### Cluster Health

I know. Why is this not the thing called "cluster-level logging"? The answer is that Google is bad at naming things. This refers to monitoring things like CPU usage, available disk space, that sort of thing. There's a [full writeup here](https://phoenixnap.com/kb/prometheus-kubernetes-monitoring), but basically it just involves pointing your log aggregator at the built-in Kubernetes monitoring endpoints to keep tabs on the Nodes and the code that's running the cluster itself. They included this nice picture.

![cluster health](https://phoenixnap.com/kb/wp-content/uploads/2020/01/prometheus-kubernetes-architechture-overview.png)

#### Generating metrics from logs

As you may have guessed, the go-to distributed logging tool is [Prometheus](https://prometheus.io/). The idea is that you can plug a bunch of data sources into it, and then it performs some log aggregation before storing results in a [time series database](https://www.influxdata.com/time-series-database/). It can be configured to delete old logs to avoid paying too much for storage, and also to delete *most* of the old logs to allow for checking metrics -- for example, you don't need to know the status of all 2,000 requests to the API from January 23, but it would be nice to know that the average response time was 1.5 seconds.

#### Tracing

Tracing is about tracking a single event as it moves through your distributed system. For example, think about what happens when a DJ wants to upload a new set.

1. DJ clicks "upload mp3" on streampusher, sending a POST request to api.datafruits.com
3. The ingress controller routes the request to the API
3. The API authenticates the request and forwards the image to object storage
4. Object storage stores the file and responds 200 OK
5. The API receives the response from object storage and responds 200 OK
6. The ingress controller forwards the response to the DJ, and streampusher says "Success!"

Even though this is one event, it touches the ingress controller, the API, and object storage. If we're trying to troubleshoot why it takes a million years to upload a set, we need to be able to trace this event. That's where something like [Jaeger](https://www.jaegertracing.io/docs/1.17/#about) comes in. This gets a little more involved, because it requires instrumenting all the code in the event lifecycle, but eventually it'd be nice to have.

#### Monitoring

Prometheus can do automated alerts (via a bunch of stuff, including [discord](https://github.com/prometheus/alertmanager/issues/1365#issuecomment-579337769)), and we can also use tools like [Grafana](https://prometheus.io/docs/visualization/grafana/) to build dashboards.

#### TL;DR

Basically, we need some kind of aggregation service to centralize our logs, and a visualization tool to make a dashboard to look at those logs.

Reply:
Wow this is certainly a lot bigger answer than I expected! :sweat_smile: 
What would an MVP of this look like, and is it still within the scope of this RFC?